### ISSUE TITLE
False positive `reportPrivateImportUsage` for `IPython.display.HTML`

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#5902]
**Describe the bug**

`IPython.display.HTML` is a public member of the `IPython.display` module, as far as I can tell anyway, but pyright is complaining that it isn't public. I believe this is a false positive because, by reading the IPython source code, `HTML` looks like it should be public:
- https://github.com/ipython/ipython/blob/bc8b2d22f3b7738a407f72e2d7dda4d531ce2bb3/IPython/display.py#L30
- https://github.com/ipython/ipython/blob/bc8b2d22f3b7738a407f72e2d7dda4d531ce2bb3/IPython/core/display.py#L27

**Code or Screenshots**

```python
# demo.py
import IPython.display

IPython.display.HTML
```

```
> pyright demo.py 
/home/kshi/pyright-bug/demo.py
  /home/kshi/pyright-bug/demo.py:3:17 - error: "HTML" is not exported from module "IPython.display" (reportPrivateImportUsage)
1 error, 0 warnings, 0 informations
```

Currently I've installed `ipython` version 8.15.0 (I think this is, as of writing, the latest version available on PyPI).

**VS Code extension or command-line**

I tested and produced this error both via the neovim LSP pyright integration and via the CLI directly.

---

For your convenience, I've also created a Dockerfile to reproduce this error in an isolated environment:

```dockerfile
FROM alpine

RUN apk update
RUN apk add npm py3-virtualenv

WORKDIR /tmp/demo

RUN npm install pyright
RUN virtualenv venv

RUN venv/bin/pip install ipython

COPY demo.py .

ENTRYPOINT ["sh"]
CMD ["-c", ". venv/bin/activate && npm exec pyright demo.py"]
```

![image](https://github.com/microsoft/pyright/assets/3887189/2807606c-6919-4455-9671-3630c6e64ebd)

**Comment by erictraut** [comment]
This isn't a false positive. Pyright is correct in flagging this as a private import. 

The `IPython` package contains a "py.typed" marker file, which means that it is assumed to follow certain rules about specifying which symbols within each submodule are intended to be public. You can read about these rules [here](https://microsoft.github.io/pyright/#/typed-libraries?id=library-interface).

The `IPython/display.py` module imports the symbol `HTML`, but it doesn't specify that it should be re-exported from this submodule. It is therefore considered private. To make it public, the maintainers of IPython would need to use a redundant form of import (`from IPython.core.display import HTML as HTML`). If you make this change manually, you will see that pyright treats the symbol as public.

If you think this symbol should be public, please consider filing a bug report in the IPython project.

**Comment by kwshi** [comment]
Thanks so much for the quick response. One question: are the rules you linked something generally agreed upon across other Python typing utilities (e.g., mypy) and documented elsewhere as well (e.g., a PEP?), or is it a Pyright-specific convention?

**Comment by hmc-cs-mdrissi** [comment]
https://typing.readthedocs.io/en/latest/source/stubs.html#imports

The rules are documented here as well. There was a typing-sig discussion on public API rules that reached agreement with other type checkers and mypy should is expected to follow same rules listed in that page.

**Comment by erictraut** [comment]
Yes, these rules are also documented [here](https://github.com/python/typing/blob/master/docs/source/libraries.rst), and we have agreement across type checkers on the behavior. I know that mypy has implemented some of the rules in this document, but there are others that are still unimplemented, such as [this one](https://github.com/python/mypy/issues/15300). The maintainers of IPython may want to test with the latest mypy if they have concerns about compatibility.
